Quickstep sample operator

rogersjeffreyl opened the pull request:
This PR is for adding  operator support for the sample operator.  For now support is available only for sample on the stored relations.

zuyu commented:
@jianqiao Please add comments on the `File Changed` tab, not a particular commit.

zuyu commented on query_optimizer/ExecutionGenerator.cpp:
```diff
@@ -358,6 +363,45 @@ bool ExecutionGenerator::convertSimpleProjection(
+      findRelationInfoOutputByPhysical(physical_sample->input());
+  DCHECK(input_relation_info != nullptr);
+
+  SampleOperator *sp = new SampleOperator(*input_relation_info->relation,
```
Rename `sp` to `sample_op`.

zuyu commented on query_optimizer/ExecutionGenerator.hpp:
```diff
@@ -211,6 +212,12 @@ class ExecutionGenerator {
    **/
   bool convertSimpleProjection(const QueryContext::scalar_group_id project_expressions_group_index,
                                std::vector<attribute_id> *attributes) const;
+  /**
```
Add a new line before the comment.

zuyu commented on query_optimizer/logical/Sample.cpp:
```diff
@@ -0,0 +1,61 @@
+
+#include <string>
+#include <vector>
+#include "query_optimizer/OptimizerTree.hpp"
```
Add a new line above.

zuyu commented on query_optimizer/logical/Sample.cpp:
```diff
@@ -0,0 +1,61 @@
+
+LogicalPtr Sample::copyWithNewChildren(
+    const std::vector<LogicalPtr> &new_children) const {
+  DCHECK_EQ(new_children.size(), children().size());
```
Exchange two `size`.

For `CHECK_EQ`, the left hand side is the expected value, and the right the checked.

rogersjeffreyl replied:
I found the same line  DCHECK_EQ(new_children.size(), children().size());
in query_optimizer/logical/Project.cpp. Should that be changed too?

zuyu replied:
Yes, but in another separate PR.

zuyu commented on query_optimizer/logical/Sample.cpp:
```diff
@@ -0,0 +1,61 @@
+    std::vector<OptimizerTreeBaseNodePtr> *non_container_child_fields,
+    std::vector<std::string> *container_child_field_names,
+    std::vector<std::vector<OptimizerTreeBaseNodePtr>> *container_child_fields)
+    const {
```
Combine this line with the above one after the `)`.

zuyu commented on query_optimizer/logical/Sample.cpp:
```diff
@@ -0,0 +1,61 @@
+    std::vector<std::string> *container_child_field_names,
+    std::vector<std::vector<OptimizerTreeBaseNodePtr>> *container_child_fields)
+    const {
+    non_container_child_field_names->push_back("input");
```
Two space indents in the function.

zuyu commented on query_optimizer/logical/Sample.hpp:
```diff
@@ -0,0 +1,128 @@
+#include <memory>
+#include <string>
+#include <vector>
+#include "query_optimizer/OptimizerTree.hpp"
```
Add a new line above.

zuyu commented on query_optimizer/logical/Sample.hpp:
```diff
@@ -0,0 +1,128 @@
+class Sample : public Logical {
+ public:
+  LogicalType getLogicalType() const override { return LogicalType::kSample; }
+  std::string getName() const override { return "SAMPLE"; }
```
Add a new line above, as well as the doxygen comment.

zuyu commented on query_optimizer/logical/Sample.hpp:
```diff
@@ -0,0 +1,128 @@
+ */
+class Sample : public Logical {
+ public:
+  LogicalType getLogicalType() const override { return LogicalType::kSample; }
```
Add the doxygen comment for this function.

zuyu commented on query_optimizer/logical/Sample.hpp:
```diff
@@ -0,0 +1,128 @@
+  std::string getName() const override { return "SAMPLE"; }
+
+  /**
+   * @return  The percentage of data to be sampled.
```
Remove the one space after `@return`, and add `@brief`.

zuyu commented on query_optimizer/logical/Sample.hpp:
```diff
@@ -0,0 +1,128 @@
+  /**
+   * @return The input operator.
+   */
+  const LogicalPtr& input() const { return input_; }
```
Move all non-overridden methods after overridden ones.

zuyu commented on query_optimizer/logical/Sample.hpp:
```diff
@@ -0,0 +1,128 @@
+      const std::vector<LogicalPtr> &new_children) const override;
+
+  std::vector<expressions::AttributeReferencePtr>
+  getOutputAttributes() const override {
```
Four space indents.

zuyu commented on query_optimizer/logical/Sample.hpp:
```diff
@@ -0,0 +1,128 @@
+  }
+
+  std::vector<expressions::AttributeReferencePtr>
+  getReferencedAttributes() const override {
```
Four space indents.

zuyu commented on query_optimizer/logical/Sample.hpp:
```diff
@@ -0,0 +1,128 @@
+
+ private:
+  Sample(const LogicalPtr &input,
+          const bool is_block_sample,
```
Remove one space.

zuyu commented on query_optimizer/physical/PhysicalType.hpp:
```diff
@@ -40,6 +40,7 @@ enum class PhysicalType {
   kInsertTuple,
   kNestedLoopsJoin,
   kSelection,
+  kSample,
```
Move it before `kSelection`.

zuyu commented on query_optimizer/physical/Sample.cpp:
```diff
@@ -0,0 +1,61 @@
+
+#include <string>
+#include <vector>
+#include "query_optimizer/OptimizerTree.hpp"
```
New line

zuyu commented on query_optimizer/physical/Sample.cpp:
```diff
@@ -0,0 +1,61 @@
+#include "query_optimizer/expressions/ExpressionUtil.hpp"
+#include "query_optimizer/expressions/NamedExpression.hpp"
+#include "utility/Cast.hpp"
+#include "glog/logging.h"
```
New line

zuyu commented on query_optimizer/physical/Sample.cpp:
```diff
@@ -0,0 +1,61 @@
+    std::vector<OptimizerTreeBaseNodePtr> *non_container_child_fields,
+    std::vector<std::string> *container_child_field_names,
+    std::vector<std::vector<OptimizerTreeBaseNodePtr>> *container_child_fields) const {
+
```
Remove this empty line.

zuyu commented on query_optimizer/logical/Sample.hpp:
```diff
@@ -0,0 +1,128 @@
+}  // namespace optimizer
+}  // namespace quickstep
+
+#endif /* QUICKSTEP_QUERY_OPTIMIZER_LOGICAL_SAMPLE_HPP_ */
```
Change to `#endif  // QUICKSTEP_QUERY_OPTIMIZER_LOGICAL_SAMPLE_HPP_ `

zuyu commented on query_optimizer/physical/Sample.hpp:
```diff
@@ -0,0 +1,138 @@
+  /**
+   * @return Input physical node.
+   */
+  const PhysicalPtr& input() const { return children()[0]; }
```
Move all non-overridden methods after overridden ones.

zuyu commented on query_optimizer/physical/Sample.hpp:
```diff
@@ -0,0 +1,138 @@
+   */
+  const PhysicalPtr& input() const { return children()[0]; }
+
+  PhysicalPtr copyWithNewChildren(
```
Doxygen comments for all methods in the file.

zuyu commented on query_optimizer/physical/Sample.hpp:
```diff
@@ -0,0 +1,138 @@
+      const bool is_block_sample,
+      const int percentage) {
+    return SamplePtr(
+     new Sample(input, is_block_sample, percentage));
```
Add three more space indents.

zuyu replied:
Indent issue here.

zuyu commented on query_optimizer/physical/Sample.hpp:
```diff
@@ -0,0 +1,138 @@
+  const bool is_block_sample_;
+  const PhysicalPtr input_;
+  std::vector<expressions::AttributeReferencePtr> output_attributes_;
+  DISALLOW_COPY_AND_ASSIGN(Sample);
```
New line.

zuyu commented on query_optimizer/physical/Sample.hpp:
```diff
@@ -0,0 +1,138 @@
+  const int percentage_;
+  const bool is_block_sample_;
+  const PhysicalPtr input_;
+  std::vector<expressions::AttributeReferencePtr> output_attributes_;
```
Unused member?

zuyu commented on query_optimizer/resolver/Resolver.cpp:
```diff
@@ -1208,7 +1209,13 @@ L::LogicalPtr Resolver::resolveTableReference(const ParseTableReference &table_r
       }
-
+      // Create a logical sample operator. Applicable only if \p the table is materialized.
+      // Sampling  from subquery table references is not supported.
```
Remove one space after `Sampling`.

zuyu commented on relational_operators/SampleOperator.cpp:
```diff
@@ -0,0 +1,123 @@
+
+namespace quickstep {
+
+  bool SampleOperator::getAllWorkOrders(
```
Remove two space indents.

zuyu commented on relational_operators/SampleOperator.cpp:
```diff
@@ -0,0 +1,123 @@
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::uniform_real_distribution<> dis(0, 1);
+  double probability = static_cast<double> (percentage_)/100;
```
Change to

```
  const double probability = static_cast<double>(percentage_) / 100;
```

zuyu commented on relational_operators/SampleOperator.cpp:
```diff
@@ -0,0 +1,123 @@
+
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::uniform_real_distribution<> dis(0, 1);
```
Rename `dis` to `distribution`.

zuyu commented on relational_operators/SampleOperator.cpp:
```diff
@@ -0,0 +1,123 @@
+      query_context->getInsertDestination(output_destination_index_);
+
+  std::random_device rd;
+  std::mt19937 gen(rd());
```
Rename `gen` to `generator`.

zuyu commented on relational_operators/SampleOperator.cpp:
```diff
@@ -0,0 +1,123 @@
+                                  storage_manager),
+                op_index_);
+        }
+      }
```
Refactor above `if-else` code to

```
  for (const block_id input_block_id : input_relation_block_ids_) {
    if (!is_block_sample_ || distribution(genator) <= probability) {
      container->addNormalWorkOrder(...);
    }
  }
```

rogersjeffreyl replied:
@zuyu It makes sense code wise. However, should we be checking for every block if the block sample flag is set?  Wouldnt it be better  to just make that is_block_sample check only once?

zuyu replied:
If it is constant, do things like this,

```
  const bool enable_sample = !is_block_sample_ || distribution(genator) <= probability;
  for (const block_id input_block_id : input_relation_block_ids_) {
    if (enable_sample) {
      container->addNormalWorkOrder(...);
    }
  }
```

rogersjeffreyl replied:
@zuyu 
the logic here is 
1) if it is a block sample then generate work orders for blocks based on a probability , otherwise generate work orders for all blocks.
so,the block sample flag is set only once and then we check blocks with probability i.e the probability checking should be within the loop(every time a new block is encountered a probability should be generated) . So am not sure if the above suggestion is semantically correct.

zuyu replied:
Then the code below should work. I've deleted the const version, which is incorrect.

```
  for (const block_id input_block_id : input_relation_block_ids_) {
    if (!is_block_sample_ || distribution(genator) <= probability) {
      container->addNormalWorkOrder(...);
    }
  }
```

zuyu replied:
Alright, you are fine here, but you need to align `new SampleWorkOrder` and `op_index_` in the `else` part.

zuyu commented on relational_operators/SampleOperator.cpp:
```diff
@@ -0,0 +1,123 @@
+                op_index_);
+            ++num_workorders_generated_;
+        }
+      }
```
Refactor the `if-else` code here as well.

zuyu commented on relational_operators/SampleOperator.cpp:
```diff
@@ -0,0 +1,123 @@
+  BlockReference block(
+      storage_manager_->getBlock(input_block_id_, input_relation_));
+
+  block->selectSample(is_block_sample_, percentage_, output_destination_);
```
Rename `selectSample` to `sample`?

zuyu commented on relational_operators/SampleOperator.hpp:
```diff
@@ -0,0 +1,177 @@
+class Predicate;
+class Scalar;
+class StorageManager;
+class WorkOrdersContainer;
```
Run `iwyu` scripts for all files changed in this PR. We don't use `Scalar` and `Predicate` at least, so remove them.

rogersjeffreyl replied:
Can you please point me to the scripts?

zuyu replied:
https://github.com/pivotalsoftware/quickstep/tree/master/third_party/iwyu

zuyu commented on relational_operators/SampleOperator.hpp:
```diff
@@ -0,0 +1,177 @@
+class StorageManager;
+class WorkOrdersContainer;
+
+
```
Remove this empty line.

zuyu commented on relational_operators/SampleOperator.hpp:
```diff
@@ -0,0 +1,177 @@
+ */
+
+/**
+ * @brief An operator which performs sample (block or tuple) over a relation.
```
Change to `@brief An operator which performs either block- or tuple-level sample over a relation.`

zuyu commented on relational_operators/SampleOperator.hpp:
```diff
@@ -0,0 +1,177 @@
+class SampleOperator : public RelationalOperator {
+ public:
+  /**
+   * @brief Constructor for Sample with the sampling percentage and type of samplng
```
Change `Sample` to `SampleOperator`.

Fix typo `samplng`.

zuyu commented on relational_operators/SampleOperator.hpp:
```diff
@@ -0,0 +1,177 @@
+   *        is fully available to the operator before it can start generating
+   *        workorders.
+   * @param is_block_sample Flag indicating whether the sample type is block or tuple
+   * @param percentage The percentage of data to be sampled
```
Add the period symbol for above two lines.

zuyu commented on relational_operators/SampleOperator.hpp:
```diff
@@ -0,0 +1,177 @@
+  SampleOperator(const CatalogRelation &input_relation,
+                 const CatalogRelation &output_relation,
+                 const QueryContext::insert_destination_id output_destination_index,
+                 bool input_relation_is_stored,
```
Mark it `const`.

zuyu commented on relational_operators/SampleOperator.hpp:
```diff
@@ -0,0 +1,177 @@
+  const CatalogRelation &output_relation_;
+  const QueryContext::insert_destination_id output_destination_index_;
+  const bool input_relation_is_stored_;
+  std::vector<block_id> input_relation_block_ids_;
```
Move non-const members after `const` ones.

zuyu commented on relational_operators/SampleOperator.hpp:
```diff
@@ -0,0 +1,177 @@
+
+ private:
+  const CatalogRelationSchema &input_relation_;
+  const CatalogRelation &output_relation_;
```
Use `CatalogRelationSchema` instead.

zuyu commented on relational_operators/SampleOperator.hpp:
```diff
@@ -0,0 +1,177 @@
+   * 
+   **/
+  SampleOperator(const CatalogRelation &input_relation,
+                 const CatalogRelation &output_relation,
```
Use `CatalogRelationSchema` instead.

rogersjeffreyl replied:
Sure.  If I make the input CatalogRelationSchema , how do i get the block ids ?

Previously this was the code:
input_relation_block_ids_(input_relation_is_stored ? input_relation.getBlocksSnapshot()
                                                       : std::vector<block_id>()),

zuyu replied:
I mean `output_relation` only.

rogersjeffreyl replied:
Thanks for the clarification

zuyu commented on storage/StorageBlock.cpp:
```diff
@@ -322,6 +322,62 @@ void StorageBlock::selectSimple(const std::vector<attribute_id> &selection,
+        needed_percentage = 100-percentage;
+        invert_bits = true;
+    }
+    number_of_tuples = sequence->length()*needed_percentage/100;
```
Add spaces between symbols.

zuyu replied:
Remove above declaration, and declare here as a const variable.

zuyu replied:
Remove above declaration, and declare here as a const variable.

zuyu replied:
Rename to `num_tuples`.

zuyu replied:
Btw, do you need the round-up the right side value?

zuyu commented on storage/StorageBlock.cpp:
```diff
@@ -322,6 +322,62 @@ void StorageBlock::selectSimple(const std::vector<attribute_id> &selection,
+    // calculate the number of 0's to be set(corresponding to the
+    // rows that will not be picked) and set those rows to 1.
+    if (percentage > 50) {
+        needed_percentage = 100-percentage;
```
Add spaces between symbols.

zuyu commented on storage/StorageBlock.cpp:
```diff
@@ -322,6 +322,62 @@ void StorageBlock::selectSimple(const std::vector<attribute_id> &selection,
+    destination->bulkInsertTuples(accessor.get());
+  } else {
+    int needed_percentage = percentage;
+    std::random_device rd;
```
Rename `rd` to `random_number_generator`.

zuyu commented on storage/StorageBlock.cpp:
```diff
@@ -322,6 +322,62 @@ void StorageBlock::selectSimple(const std::vector<attribute_id> &selection,
+    accessor.reset(tuple_store_->createValueAccessor());
+    destination->bulkInsertTuples(accessor.get());
+  } else {
+    int needed_percentage = percentage;
```
Rename to `actual_percentage`?

zuyu commented on storage/StorageBlock.cpp:
```diff
@@ -322,6 +322,62 @@ void StorageBlock::selectSimple(const std::vector<attribute_id> &selection,
+  } else {
+    int needed_percentage = percentage;
+    std::random_device rd;
+    std::mt19937 gen(rd());
```
Rename to `generator`?

zuyu commented on storage/StorageBlock.cpp:
```diff
@@ -322,6 +322,62 @@ void StorageBlock::selectSimple(const std::vector<attribute_id> &selection,
+    std::unique_ptr<TupleIdSequence> sequence_mask(
+        new TupleIdSequence(sequence->length()));
+    std::unordered_map<std::size_t, std::size_t> tuple_index_mapping;
+    std::size_t number_of_tuples = 0;
```
Rename to `num_tuples`.

zuyu commented on storage/StorageBlock.cpp:
```diff
@@ -322,6 +322,62 @@ void StorageBlock::selectSimple(const std::vector<attribute_id> &selection,
+        } else {
+            sequence_mask->set(tuple_index_mapping[random_number], true);
+        }
+        tuple_index_mapping[random_number] = sequence->length()-(n+1);
```
Two space indents here.

zuyu commented on storage/StorageBlock.cpp:
```diff
@@ -322,6 +322,62 @@ void StorageBlock::selectSimple(const std::vector<attribute_id> &selection,
+    // TupleIdSequence
+    // Generate a random number between 0 and N
+    for (std::size_t n = 0; n < number_of_tuples; n++) {
+        std::uniform_real_distribution<> dis(0, sequence->length() - (n+1));
```
Rename to `distribution`.

Add spaces between symbols.

zuyu commented on storage/StorageBlock.cpp:
```diff
@@ -322,6 +322,62 @@ void StorageBlock::selectSimple(const std::vector<attribute_id> &selection,
+    // Generate a random number between 0 and N
+    for (std::size_t n = 0; n < number_of_tuples; n++) {
+        std::uniform_real_distribution<> dis(0, sequence->length() - (n+1));
+        std::size_t random_number = dis(gen);
```
Mark it `const`.

zuyu commented on storage/StorageBlock.cpp:
```diff
@@ -322,6 +322,62 @@ void StorageBlock::selectSimple(const std::vector<attribute_id> &selection,
+        if (tuple_index_mapping.find(random_number) == tuple_index_mapping.end()) {
+          sequence_mask->set(random_number, true);
+        } else {
+            sequence_mask->set(tuple_index_mapping[random_number], true);
```
Two space indent.

zuyu commented on storage/StorageBlock.cpp:
```diff
@@ -322,6 +322,62 @@ void StorageBlock::selectSimple(const std::vector<attribute_id> &selection,
+          sequence_mask->set(random_number, true);
+        } else {
+            sequence_mask->set(tuple_index_mapping[random_number], true);
+        }
```
Prefer to use `? : ` to refactor above `if-else` code.

rogersjeffreyl commented:
@zuyu I have addressed the review comments.. Kindly let me know if it looks fine.

zuyu commented:
@rogersjeffreyl You need to pass the CI first!

zuyu commented:
@rogersjeffreyl Reset `/build/qsstor/catalog.pb.bin` first.

zuyu commented on query_optimizer/ExecutionGenerator.cpp:
```diff
@@ -358,6 +363,45 @@ bool ExecutionGenerator::convertSimpleProjection(
+                                          insert_destination_index,
+                                          input_relation_info->isStoredRelation(),
+                                          physical_sample->is_block_sample(),
+                                          physical_sample->percentage());
```
Align the above line with `(`.

zuyu commented on query_optimizer/ExecutionGenerator.hpp:
```diff
@@ -212,6 +213,13 @@ class ExecutionGenerator {
+   *
+   * @param physical_plan The Sample to be converted.
+   */
+  void convertSample(const physical::SamplePtr &physical_plan);
```
Move it above `convertSimpleProjection` and even `convertSelection`.

zuyu commented on query_optimizer/logical/Sample.cpp:
```diff
@@ -0,0 +1,61 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *   University of Wisconsin—Madison.
```
Add four space indents.

zuyu replied:
Two whitespace needed.

zuyu commented on query_optimizer/logical/Sample.cpp:
```diff
@@ -0,0 +1,61 @@
+    std::vector<std::string> *non_container_child_field_names,
+    std::vector<OptimizerTreeBaseNodePtr> *non_container_child_fields,
+    std::vector<std::string> *container_child_field_names,
+    std::vector<std::vector<OptimizerTreeBaseNodePtr>> *container_child_fields)const {
```
Add a whitespace before `const`.

zuyu commented on query_optimizer/logical/Sample.hpp:
```diff
@@ -0,0 +1,133 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *   University of Wisconsin—Madison.
```
Add four whitespace indents.

rogersjeffreyl replied:
This was copy pasted from another file. Most of the copy right text in the other files follows the same formatting

zuyu replied:
The existing files might still have leftover style issues, which should be fixed in a separate PR.

For new files, we should stick to the style guidance.

zuyu replied:
Two space indents is fine.

zuyu commented on query_optimizer/logical/Sample.hpp:
```diff
@@ -0,0 +1,133 @@
+  Sample(const LogicalPtr &input,
+         const bool is_block_sample,
+         const int percentage)
+       : input_(input),
```
I believe the indent before `:` is more than four-whitespace.

zuyu commented on query_optimizer/logical/Sample.hpp:
```diff
@@ -0,0 +1,133 @@
+       : input_(input),
+         is_block_sample_(is_block_sample),
+         percentage_(percentage) {
+     addChild(input);
```
Reduce one whitespace.

zuyu commented on query_optimizer/logical/Sample.hpp:
```diff
@@ -0,0 +1,133 @@
+}  // namespace optimizer
+}  // namespace quickstep
+
+#endif  // QUICKSTEP_QUERY_OPTIMIZER_LOGICAL_SAMPLE_HPP_  //
```
Remove the last `//`.

zuyu commented on query_optimizer/physical/Sample.cpp:
```diff
@@ -0,0 +1,62 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *   University of Wisconsin—Madison.
```
Add four whitespace indents.

rogersjeffreyl replied:
do you mean between the  * and the text?

zuyu replied:
Yes, before `University`.

rogersjeffreyl replied:
This was copy pasted from another file. Most of the copy right text in the other files follows the same formatting

zuyu replied:
Need two whitespace at least for a part of sentence.

zuyu commented on query_optimizer/physical/Sample.hpp:
```diff
@@ -0,0 +1,138 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *   University of Wisconsin—Madison.
```
Add four whitespace indents.

zuyu commented on query_optimizer/physical/Sample.hpp:
```diff
@@ -0,0 +1,138 @@
+}  // namespace optimizer
+}  // namespace quickstep
+
+#endif /* QUICKSTEP_QUERY_OPTIMIZER_SAMPLE_HPP_ */
```
Change to `#endif  // QUICKSTEP_QUERY_OPTIMIZER_SAMPLE_HPP_`

zuyu commented on query_optimizer/strategy/CMakeLists.txt:
```diff
@@ -96,6 +97,7 @@ target_link_libraries(quickstep_queryoptimizer_strategy_OneToOne
                       quickstep_queryoptimizer_physical_InsertTuple
                       quickstep_queryoptimizer_physical_Physical
                       quickstep_queryoptimizer_physical_SharedSubplanReference
+                      quickstep_queryoptimizer_physical_Sample
```
Move this line above.

zuyu commented on relational_operators/SampleOperator.cpp:
```diff
@@ -0,0 +1,121 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *   University of Wisconsin—Madison.
```
Add four whitespace indents.

rogersjeffreyl replied:
This was copy pasted from another file. Most of the copy right text in the other files follows the same formatting

zuyu commented on relational_operators/SampleOperator.cpp:
```diff
@@ -0,0 +1,121 @@
+  std::random_device random_device;
+  std::mt19937 generator(random_device());
+  std::uniform_real_distribution<> distribution(0, 1);
+  double probability = static_cast<double> (percentage_) / 100;
```
Mark it `const`.

zuyu commented on relational_operators/SampleOperator.cpp:
```diff
@@ -0,0 +1,121 @@
+          while (num_workorders_generated_ < input_relation_block_ids_.size()) {
+            container->addNormalWorkOrder(
+                new SampleWorkOrder(
+                                    input_relation_,
```
Move this line connecting with the above one.

zuyu commented on relational_operators/SampleOperator.cpp:
```diff
@@ -0,0 +1,121 @@
+                                    output_destination,
+                                    storage_manager),
+                op_index_);
+            ++num_workorders_generated_;
```
Remove two whitespace indents.

zuyu commented on relational_operators/SampleOperator.hpp:
```diff
@@ -0,0 +1,174 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *   University of Wisconsin—Madison.
```
Add four whitespace indents.

zuyu commented on storage/StorageBlock.cpp:
```diff
@@ -322,6 +322,60 @@ void StorageBlock::selectSimple(const std::vector<attribute_id> &selection,
 
+void StorageBlock::sample(const bool is_block_sample,
+                                const int percentage,
+                                InsertDestinationInterface *destination) const {
```
Align with `(` above.

zuyu commented on storage/StorageBlock.cpp:
```diff
@@ -322,6 +322,60 @@ void StorageBlock::selectSimple(const std::vector<attribute_id> &selection,
+    std::unique_ptr<TupleIdSequence> sequence_mask(
+        new TupleIdSequence(sequence->length()));
+    std::unordered_map<std::size_t, std::size_t> tuple_index_mapping;
+    std::size_t num_tuples = 0;
```
1) Delete this one.

rogersjeffreyl replied:
@zuyu  which one should be deleted?

zuyu replied:
Delete this line, and see `2)` below for more info.

zuyu commented on storage/StorageBlock.cpp:
```diff
@@ -322,6 +322,60 @@ void StorageBlock::selectSimple(const std::vector<attribute_id> &selection,
+        actual_percentage = 100 - percentage;
+        invert_bits = true;
+    }
+    num_tuples = sequence->length() * actual_percentage / 100;
```
2) Add `const std::size_t `.

zuyu commented on storage/StorageBlock.cpp:
```diff
@@ -322,6 +322,60 @@ void StorageBlock::selectSimple(const std::vector<attribute_id> &selection,
+    // rows that will not be picked) and set those rows to 1.
+    if (percentage > 50) {
+        actual_percentage = 100 - percentage;
+        invert_bits = true;
```
Change to two whitespace indents.

zuyu commented on storage/StorageBlock.cpp:
```diff
@@ -322,6 +322,60 @@ void StorageBlock::selectSimple(const std::vector<attribute_id> &selection,
+        std::size_t random_number = distribution(generator);
+        tuple_index_mapping.find(random_number) == tuple_index_mapping.end()
+            ? sequence_mask->set(random_number, true)
+            : sequence_mask->set(tuple_index_mapping[random_number], true);
```
Well, I mean

```
sequence_mask->set(tuple_index_mapping.find(random_number) == tuple_index_mapping.end()
                       ? random_number
                       : tuple_index_mapping[random_number],
                   true);
```

zuyu commented on storage/StorageBlock.cpp:
```diff
@@ -322,6 +322,60 @@ void StorageBlock::selectSimple(const std::vector<attribute_id> &selection,
+    // TupleIdSequence
+    // Generate a random number between 0 and N
+    for (std::size_t n = 0; n < num_tuples; n++) {
+        std::uniform_real_distribution<> distribution(0, sequence->length() - (n+1));
```
Change to `(n + 1)`.

zuyu commented on storage/StorageBlock.cpp:
```diff
@@ -322,6 +322,60 @@ void StorageBlock::selectSimple(const std::vector<attribute_id> &selection,
+        tuple_index_mapping.find(random_number) == tuple_index_mapping.end()
+            ? sequence_mask->set(random_number, true)
+            : sequence_mask->set(tuple_index_mapping[random_number], true);
+        tuple_index_mapping[random_number] = sequence->length()-(n+1);
```
Add whitespace between the right hand side expressions.

zuyu commented on storage/StorageBlock.hpp:
```diff
@@ -334,6 +334,24 @@ class StorageBlock : public StorageBlockBase {
+  void sample(const bool is_block_sample,
+              const int percentage,
+              InsertDestinationInterface *destination) const;
+  /**
```
Add a new line above.

rogersjeffreyl replied:
@zuyu  you mean between the end of comments and beginning of the function?
I had a look at the file: the previously written code has  the following structure:
<newline>
/**
 * comment for the function
*
*/ 
function
 <newline>
/* comment for the new function

is the existing styling wrong?

zuyu replied:
@rogersjeffreyl No, I mean the end of the function (Line 353) and the beginning of the comment (Line 354).

zuyu commented on storage/StorageBlock.hpp:
```diff
@@ -334,6 +334,24 @@ class StorageBlock : public StorageBlockBase {
+   *         sampling percentage. In the case of block sample the entire
+   *         block is returned.
+   **/
+  void sample(const bool is_block_sample,
```
Move this function declaration before `select`.

zuyu commented on storage/StorageBlock.cpp:
```diff
@@ -322,6 +322,60 @@ void StorageBlock::selectSimple(const std::vector<attribute_id> &selection,
                                                       accessor.get());
 }
 
+void StorageBlock::sample(const bool is_block_sample,
```
Move this function definition before `select`.

rogersjeffreyl commented:
@zuyu have addressed your comments.  For the copyright text. I have modified logical/Sample.hpp. Please let me know if that looks fine and  I will do the same for the other files as well

zuyu commented:
@rogersjeffreyl This PR looks good, except for minor style issues:
 1. we do need two whitespace indents in the copyright comment in the line 2 before `University`. I do not know which file you are referring to, but check out https://github.com/pivotalsoftware/quickstep/blob/master/transaction/StronglyConnectedComponents.hpp.
 2. This line has indent issues as well: https://github.com/pivotalsoftware/quickstep/blob/quickstep_sample_operator/query_optimizer/physical/Sample.hpp#L102

zuyu commented:
@rogersjeffreyl Look good to me.

Btw, could you please squash all the commits into one?
